### PR TITLE
Validate emulator download by checking file size

### PR DESCRIPTION
**Pull request by ryanpbrewster**
There are still customer reports of corrupted downloads that finish successfully. This is a fairly blunt way of double-checking that the download is valid.

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
- computing a hash requires pulling in a new dependency (the `crypto` package)
- almost every developer will understand file size, so the error message may help more
- I am very lazy

File: src/emulator/download.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ module.exports = (name) => {
     let tmpFile = tmp.fileSync();
     let req = request.get(emulator.remoteUrl);
     let writeStream = fs.createWriteStream(tmpFile.name);
+    let bytesDownloaded = 0;
```
Comment (bkendall) [inline src/emulator/download.js]: doesn't look like this is used

**Comment by bkendall** [comment]
> * computing a hash requires pulling in a new dependency (the `crypto` package)

`crypto` is a built-in library of `node`, so it's not actually much heavy lifting. that said, that's fine. If you start getting errors where it's corrupted but the right size, you may want to consider using a hash...

**Comment by ryanpbrewster** [comment]
Went ahead and added a checksum as well. Code is now a bit messy, would appreciate pointers here.

File: src/emulator/download.js [hunk 2]
```diff
@@ -35,9 +36,24 @@ module.exports = (name) => {
...
+        stream.on('end', function() {
+          const checksum = hash.digest('hex');
+          if (checksum != emulator.expectedHash) {
+            reject(
+              new FirebaseError(
+                `download failed, expected checksum ${emulator.expectedHash} but got ${checksum}`,
+                { exit: 1 }
+              )
+            );
+          } else {
```
Comment (bkendall) [inline src/emulator/download.js]: this else block can be removed if the `reject` above is returned

File: src/emulator/download.js [hunk 3]
```diff
@@ -35,9 +36,24 @@ module.exports = (name) => {
           )
         );
       } else {
-        fs.copySync(tmpFile.name, emulator.localPath);
-        fs.chmodSync(emulator.localPath, 0o755);
-        resolve();
+        const hash = crypto.createHash('md5'),
+        const stream = fs.createReadStream(tmpFile.name);
+        stream.on('data', data => hash.update(data));
+        stream.on('end', function() {
```
Comment (bkendall) [inline src/emulator/download.js]: this can be an arrow function

File: src/emulator/download.js [hunk 4]
```diff
@@ -26,9 +27,34 @@ module.exports = (name) => {
...
+            { exit: 1 }
+          )
+        );
+      } else {
+        const hash = crypto.createHash('md5'),
+        const stream = fs.createReadStream(tmpFile.name);
+        stream.on('data', data => hash.update(data));
+        stream.on('end', function() {
+          const checksum = hash.digest('hex');
+          if (checksum != emulator.expectedHash) {
```
Comment (bkendall) [inline src/emulator/download.js]:
```suggestion
          if (checksum !== emulator.expectedHash) {
```

File: src/emulator/download.js (hunk 2 above)
Comment (bkendall) [reply src/emulator/download.js]: (and the same is true of the `else` block that _this_ entire check is in)

File: src/emulator/download.js (hunk 3 above)
Comment (ryanpbrewster) [reply src/emulator/download.js]: done

File: src/emulator/download.js (hunk 4 above)
Comment (ryanpbrewster) [reply src/emulator/download.js]: #javascript

File: src/emulator/download.js (hunk 2 above)
Comment (ryanpbrewster) [reply src/emulator/download.js]: converted these to ternary statements, lmk if that's non-idiomatic

**Comment by bkendall** [comment]
LGTM. fix up the CI and you're good to go.

**Pull request merged**